feat(integrations): AWS Lambda docs (#2937)
getsentry/sentry-docs: src/docs/product/integrations/aws-lambda/index.mdx

#### Error `SentryRole already exists`

The stack already exists on your account; you can proceed to the next step of adding your AWS account information.

#### Error `Invalid existing layer`

Your Lambda function has a layer that no longer exists. Please remove the problematic layer, then enable Sentry on the Lambda in the configuration view for the integration.

#### Error `Please validate the CloudFormation stack was created successfully`

Sentry is unable to access your AWS account through your CloudFormation stack. This means either the stack has not been created or the external ID used doesn't match the one in your installation. You can go to the Parameters tab of the CloudFormation stack to copy the value of the external ID and paste it into the form field.

## Configure

The integration configuration view allows you to do the following with your Lambda functions:

- Enable Sentry
- Disable Sentry
- Update the Sentry Lambda Layer to the latest version

In the table view, you will also see the version of the Lambda layer below the name.

![](configure.png)

## Remove Installation

> Note that when the integration is uninstalled, your Lambda functions will still be instrumented with Sentry and you will continue to receive errors. If you want to disable Sentry for all your Lambda functions, you should disable Sentry for each function before uninstalling. Note, you can also uninstall and then reinstall if you would like to change the Sentry project for your Lambdas.

Reviewer: we have a couple of Notes here, which makes this a bit hard to parse
```suggestion
When the integration is uninstalled, your Lambda functions will still be instrumented with Sentry and you will continue to receive errors. If you want to disable Sentry for all your Lambda functions, you should disable Sentry for each function before uninstalling. Note that you can also uninstall, then reinstall if you would like to change the Sentry project for your Lambdas.
```

---
title: AWS Lambda (Node)
sidebar_order: 1
description: "Learn more about Sentry's AWS Lambda integration and how you can automatically instrument your Node Lambda functions with Sentry without changing your code."
---

<Alert title="Note" level="info">

> Looking for instructions to set up the Node serverless SDK manually? [Check out these docs instead.](/platforms/node/guides/aws-lambda/)

Reviewer: nit
```suggestion
Looking for instructions to set up the Node serverless SDK manually? [Check out these docs instead](/platforms/node/guides/aws-lambda/).
```

</Alert>

> Connect Sentry to your AWS account to automatically add Sentry error monitoring and transactions to your Node Lambda functions.

Reviewer: ```suggestion
Connect Sentry to your AWS account to automatically add Sentry error and performance monitoring to your Node Lambda functions.
```

## Install

1. Navigate to **Settings > Integrations > AWS Lambda**.

   ![](directory.png)

> 1. Press **Add Installation** which will open the installation modal.

Reviewer: ```suggestion
1. Press **Add Installation** to open the installation modal.
```

   ![](integration_detail.png)

> 1. Select a project to use for your Lambda functions. Note that this project cannot be changed later. If you only have a single project, this step is skipped.

Reviewer: Can we better explain what we mean by "this project cannot be changed later" - does that mean if you select the wrong project, you can't just go back, you'd need to uninstall, then start the process over with a new project?
```suggestion
1. Select the project for your Lambda integration. **Note:** This project cannot be changed later. If you only have a single project, you'll skip this step.
```

   ![](project.png)

> 1. Click on the link that says **Configure AWS**.

Reviewer 2: Nit pick, should this be "Click on the button that says Configure AWS"?

> 1. Click on the button that says **Configure AWS**.

Reviewer: ```suggestion
1. Click to **Configure AWS**.
```

Author: @PeloWriter What about `Click on **Configure AWS**`? I feel like `Click to` is a weird thing to say.

Reviewer: Sure. My goal was simply to use the button language as part of the construction of the sentence, but I don't feel strongly about it.

   ![](aws_account.png)
  

> 1. Scroll to the bottom of the create stack page, mark the checkbox at the bottom and press the **Create Stack** button. This will create a CloudFormation stack in the region you are currently in. This stack allows Sentry to instrument your Lambda functions automatically.

Reviewer: ```suggestion
1. Scroll to the bottom of the Create Stack page, check the box, then press **Create Stack** to create a CloudFormation stack in your current region. This stack allows Sentry to automatically instrument your Lambda functions.
```

> 1. Scroll to the bottom of the create stack page, mark the checkbox at the bottom and press the **Create Stack** button. This will create a CloudFormation stack in the region you are currently in. This stack gives Sentry write access to your Lambda functions so we can automatically instrument them.

Reviewer 2: I will reword to "This stack allows Sentry to instrument your Lambda functions automatically"

   ![](cloudformation.png)

> 1. Wait a minute or two for the stack to be created. When you see the status change to **Create Complete**, copy your AWS account number, select the region you want to use, and finally press **Next**.

Reviewer 2: "press Next on the installation modal"

> 1. Wait a minute or two for the stack to be created. When you see the status change to **Create Complete**, copy your AWS account number, select the region you want to use, and finally press **Next** in the installation modal.

Reviewer: ```suggestion
1. Wait for the stack to be created, which is confirmed by seeing the status change to **Create Complete**, then copy your AWS account number, select the region you want to use, and press **Next**.
```

   ![](stack_created.png)
   
<Alert title="Note" level="info">

> You only need to create the CloudFormation stack once per account even when installing the integration on multiple regions. If you are setting up the integration on a second region, verify that the externalId in the form matches the one for the existing CloudFormation stack on the "Parameters" tab.

Reviewer: ```suggestion
You need to create the CloudFormation stack only once per account, even when installing the integration in multiple regions. If you are setting up the integration in a second region, verify that the externalId in the form matches the one for the existing CloudFormation stack on the "Parameters" tab.
```

</Alert>

   

> 1. Review the Lambda functions you want instrumented with Sentry. Lambda functions which are selected will get the Sentry layer and necessary environment variables automatically added when you press **Finish Setup**. Note that you can enable/disable Sentry on Lambda functions after installation as well.

Reviewer: ```suggestion
1. Select the Lambda functions to automatically add the Sentry layer and necessary environment variables when you **Finish Setup**. **Note:** You can enable or disable Sentry on Lambda functions after installation.
```

   ![](select_lambdas.png)

> 1. Depending on how many Lambda functions you enable, you might need to wait a few minutes for this step to complete. If any functions failed to instrument, follow the troubleshooting guide below.

Reviewer: This isn't a step, unless we simply want to say "Wait" - otherwise, I think what we mean to say is
```suggestion
Wait a few minutes as the Lambda functions are enabled. If any functions failed to instrument, follow the suggestions in Troubleshooting.
```

   
### Troubleshooting

> 

Reviewer: Add a sentence or two here. Perhaps something like "While uncommon, the following error messages point to specific failures. Use this list to troubleshoot and successfully instrument your Lambda functions." We should also suggest what the process would be if there issue doesn't appear below (paying customers need to contact support, non-paying can go to our forums).

Author: @PeloWriter Got it. Could you give me a blurb about contacting support? I'm guessing we have something like this somewhere else that I could use.

Reviewer: Sure. You can modify this to suit 

If you need help solving issues with your Sentry JavaScript SDK integration, you can read the edge cases documented here. If you need additional help, you can view our forums, and customers on a paid plan may also contact support.

> 1. Depending on how many Lambda functions you enable, you might need to wait a few minutes for this step to complete. If Sentry is able to update all you functions successfully, installation is considered finished. If there are errors, you will see a list of which Lambda functions failed along with a description on why they failed. 

Reviewer 2: We can reduce words here. Maybe something like: "That's it. <br> If any functions failed to instrument, follow the troubleshooting  guide below"

#### Error `SentryRole already exists`

> The stack already exists on your account so you can go to the next step of adding your AWS account information.

Reviewer: ```suggestion
The stack already exists on your account; you can proceed to the next step of adding your AWS account information.
```

#### Error `Invalid existing layer`

> Your Lambda function has a layer that no longer exists. Please remove the problematic layer and then enable Sentry on the Lambda in the configuration view for the integration.

Reviewer: ```suggestion
Your Lambda function has a layer that no longer exists. Please remove the problematic layer, then enable Sentry on the Lambda in the configuration view for the integration.
```

#### Error `Please validate the CloudFormation stack was created successfully` 

> Sentry is unable to access your AWS account through your CloudFormation stack. This means either the stack has not been created or the external ID used doesn't match the one in your installation. You can go to the Parameters tab of the CloudFormation stack to get the value of the external ID and paste it into the form field.

Reviewer: ```suggestion
Sentry is unable to access your AWS account through your CloudFormation stack. This means either the stack has not been created or the external ID used doesn't match the one in your installation. You can go to the Parameters tab of the CloudFormation stack to copy the value of the external ID and paste it into the form field.
```

## Configure

The integration configuration view allows you to do the following with your Lambda functions:

* Enable Sentry
* Disable Sentry
* Update the Sentry Lambda Layer to the latest version

In the table view, you will also see the version of the Lambda layer below the name.

   ![](configure.png)
   
   

> ## Uninstallation

Reviewer 2: This seems confusing.

Author: @ajjindal what's confusing about this? What should I change?

Reviewer: How about we change the title to "Remove Installation"

We should be explicit about how/why/when to do this (if, for example, you chose the wrong project at the start)

Note that when the integration is uninstalled, your Lambda functions will still be instrumented with Sentry and you will continue to receive errors. If you want to disable Sentry for all your Lambda functions, you should disable Sentry for each function before uninstalling.

> ## How it Works

Reviewer: This section should moved to a separate page. In the current structure, it's explaining something that's already done (so not useful at this juncture) and is reference content (whereas this page is task-focused content)

Author: @PeloWriter what page should I put this then? I feel like this is useful for people who want to know more information about implementation but I don't' want to make the installation docs even longer.

Reviewer: I don't see another integration that includes this, so there isn't a current model, but it's good content. Just make it a subpage.

When Sentry is added to a Lambda function, the following modifications are made to your Lambda functions:
* The layer for Sentry is added to your Lambdas. Any existing layers are honored and the Sentry layer is appended to the end.

   ![](layer.png)
   
* Sentry adds the following environment variables
  * `NODE_OPTIONS`: This is to preload the `awslambda-auto` module which will automatically initialize Sentry
  * `SENTRY_DSN`: This is set to the DSN of your project

>   * `SENTRY_TRACES_SAMPLE_RATE`: This sets the sampling rate for transactions

Reviewer 2: We should explain that users can adjust this parameter to control the sample rate for each Lambda function. Or may be just point to perf docs for more info.
